### DISCUSSION TITLE
License discussion

### DISCUSSION THREAD

**Discussion by CraftedRO** [TrinityCore/TrinityCore#27245] [General]
License discussion

**Comment by jackpoz** [comment]
As mentioned already at https://github.com/TrinityCore/TrinityCore/issues/27229#issuecomment-956174832 :
>AzerothCore license AGPL is incompatible with TrinityCore GPL so we cannot accept any code copied from them (unless the original author decides to submit it, then "maybe" that's ok).

You can read more about the topic at https://www.gnu.org/licenses/license-list.en.html

>Please note that the GNU AGPL is not compatible with GPLv2. It is also technically not compatible with GPLv3 in a strict sense: you cannot take code released under the GNU AGPL and convey or modify it however you like under the terms of GPLv3, or vice versa. However, you are allowed to combine separate modules or source files released under both of those licenses in a single project, which will provide many programmers with all the permission they need to make the programs they want. See section 13 of both licenses for details.

Including AC code will mean some files in TC will be released under AGPL license, forcing every TC user to make their source code available. That's not what we want to do right now so please avoid copying code from projects with a license other than AGPL.

If you want to discuss further we can take it to https://github.com/TrinityCore/TrinityCore/discussions which seems to be a more fitting place than an issue 😃

**Comment by CraftedRO** [comment]
It will be commited anyway , no point in closing this , is not a part of the core code .

**Comment by CraftedRO** [comment]
Btw if you are bored enough to close valid issues , take a look at https://github.com/TrinityCore/TrinityCore/issues/26916 that you broke it when updating to myql 5.7

**Comment by jackpoz** [comment]
> It will be commited anyway , no point in closing this , is not a part of the core code .

SQL is still code and licenses apply to that too.

**Comment by jackpoz** [comment]
> Btw if you are bored enough to close valid issues

Well now you are just lashing out at me 😢

**Comment by CraftedRO** [comment]
I just wanted to be fair and point the source where I get that update , otherwise I coud lie and don't link it , woud this be better ?

**Comment by jackpoz** [comment]
No, the "better" way would be to fix it ourselves without copying code from other projects, especially if their license is not compatible.

It's good you shared the code and I tried to write a long enough explaination instead of just closing the issue and locking it.

**Comment by Aokromes** [comment]
Offle already have this fix (and few more) without having to credit AC.

**Comment by CraftedRO** [comment]
> No, the "better" way would be to fix it ourselves without copying code from other projects, especially if their license is not compatible.
> 
> It's good you shared the code and I tried to write a long enough explaination instead of just closing the issue and locking it.

How can be proved that I fixed myself or I just copied it from other project ?

**Reply by ZaDarkSide** [reply]
Somebody will call you out when they find you copy/pasted from another project soon or later :D

**Reply by CraftedRO** [reply]
And if I really fixed too same thing , It is not possible ?

**Comment by jackpoz** [comment]
In general it would be better if the original author submitted sql fixes or c++ code instead of someone else copying his work and sending it to TC. We have had some C++ pull request from people that didn't know C++ at all so they were not able to fix the issues we spotted during the pull request review.

**Comment by acidmanifesto** [comment]
Lets keep it professions @CraftedRO this isnt high school social media drama or what ever teens do with tide pods these days.

**Comment by Faq** [comment]
The original author (code owner) could be allowing to use his fix by commenting in that issue/PR what is the same as putting signature on code.
I guess I'm not the only one who checks someone's code and comes up with idea to fix issue in another way or improve the code by putting some part from myself (in commit still can give credit to original author for the used idea).

Btw github allows convert issue to discussion.

**Comment by Teppic1** [comment]
Regarding the GPL2/3 stuff, any changes to the code (even if it entirely replaces sections), are automatically distributed under the same licence. That's basically the entire point of the GPL, it guarantees that all modifications always remain under the GPL. Since Sunwell was based on TC, it is under the GPL, and any code based off Sunwell is also under the GPL.

Nobody can change those rights or change the licence. If you don't agree to the GPL, you're not allowed to modify the code or distribute it in any way.

**Reply by jackpoz** [reply]
> Since Sunwell was based on TC, it is under the GPL

You are missing the step where Sunwell didn't release their code and as long as they didn't release any binaries they were not required to release their private changes (which GPL perfectly allows). So even if the code has a GPL license attached, it was actually never released but leaked, so we cannot use either any code from Sunwell.

**Reply by ihm-tswow** [reply]
Sunwell was not leaked, xinef released it himself, but left many code files without license headers. The new sunwell.pl is not at all affiliated with the old server. Either way, someone breaching your license does not automatically mean your code is up for grabs for any GPL project that wants it.

The release page sunwell dropped. It's currently hosted by azerothcore, but I can go dig up the original page in the archives.
https://www.azerothcore.org/pages/sunwell.pl/

**Reply by Teppic1** [reply]
Yeah, current Sunwell is based on the public old Sunwell code, which wasn't leaked but released normally. Though I expect current Sunwell is really just running AzerothCore now, which was also based on the old released code.

As for leaking, if you're given access to GPL code you have every right to distribute it under the GPL in any way you want, but a hacker for instance hasn't received the the code via distribution and so a leak that way would just be stolen code, not licensed.

**Reply by jackpoz** [reply]
I always heard it was not released, my bad, I guess that changes things when strictly speaking about Sunwell fork

**Comment by Jinnaix** [comment]
> Regarding the GPL2/3 stuff, any changes to the code (even if it entirely replaces sections), are automatically distributed under the same licence. That's basically the entire point of the GPL, it guarantees that all modifications always remain under the GPL. Since Sunwell was based on TC, it is under the GPL, and any code based off Sunwell is also under the GPL.
> 
> Nobody can change those rights or change the licence. If you don't agree to the GPL, you're not allowed to modify the code or distribute it in any way.

Azerothcore did not change the whole code to a new license. But everything commited  to AC after 17. August 2016 is AGPL.
See 
https://github.com/azerothcore/azerothcore-wotlk/blob/master/LICENSE-AGPL3
and
https://github.com/azerothcore/azerothcore-wotlk/issues/21

Further down, AGPL is incompatible with Trinitycores GPLv2

>GNU Affero General Public License (AGPL) version 3 (#AGPL) (#AGPLv3.0)
This is a free software, copyleft license. Its terms effectively consist of the terms of GPLv3, with an additional paragraph in section 13 to allow users who interact with the licensed software over a network to receive the source for that program. We recommend that developers consider using the GNU AGPL for any software which will commonly be run over a network.

>Please note that the GNU AGPL is not compatible with GPLv2. It is also technically not compatible with GPLv3 in a strict sense: you cannot take code released under the GNU AGPL and convey or modify it however you like under the terms of GPLv3, or vice versa. However, you are allowed to combine separate modules or source files released under both of those licenses in a single project, which will provide many programmers with all the permission they need to make the programs they want. See section 13 of both licenses for details.

Source: https://www.gnu.org/licenses/license-list.en.html

**Comment by ihm-tswow** [comment]
> Regarding the GPL2/3 stuff, any changes to the code (even if it entirely replaces sections), are automatically distributed under the same licence. That's basically the entire point of the GPL, it guarantees that all modifications always remain under the GPL. Since Sunwell was based on TC, it is under the GPL, and any code based off Sunwell is also under the GPL.
> 
> Nobody can change those rights or change the licence. If you don't agree to the GPL, you're not allowed to modify the code or distribute it in any way.

This is wrong. New code is not automatically distributed under the same license, it's just combined with the other code without permission. Furthermore, TrinityCore uses the "GPLv2-or-later" clause, meaning it permits derivatives with other (more restrictive) versions of the GPL as long as it's published by the fsf. The AGPL is published by the fsf, and it's thus possible to redistribute trinitycores source code with patches released under the AGPL. The opposite is not possible, any code you pulled back would retain its AGPL status and you could not revert it back to GPLv2 OL. 

However, you need to read the specific headers for individual files that azerothcore publish, because not all of their code and patches are AGPL. There was a huge PR in which they changed the language in lots of their files (on my advice) from "gplv2" to "gplv2-or-later" (because they typo'd it and accidentally made their own code incompatible with their own code). You can just read this pr for files that are still licensed gplv2-or-later.

**Comment by ihm-tswow** [comment]
here's that PR https://github.com/azerothcore/azerothcore-wotlk/pull/4762

**Comment by Teppic1** [comment]
You can read the licence at

https://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html

It states very clearly that you can only distribute changes to the code if you supply those changes under the same licence.

From the FAQ

"The GPL does not require you to release your modified version, or any part of it. You are free to make modifications and use them privately, without ever releasing them. This applies to organizations (including companies), too; an organization can make a modified version and use it internally without ever releasing it outside the organization.

**But if you release the modified version to the public in some way, the GPL requires you to make the modified source code available to the program's users, under the GPL.**

....

Section 2 says that modified versions you distribute must be licensed to all third parties under the GPL. “All third parties” means absolutely everyone—but this does not require you to do anything physically for them. It only means they have a license from you, under the GPL, for your version."

**Comment by ihm-tswow** [comment]
It must **comply** with the GPL, but **additions** or combined works need not be GPL (you're allowed to combine MIT and GPL code, for example). There are _some_ complications about changing things like code in individual files (i.e. is it a new module if I add a new function to an already gplv2-ol file?), and this is why azerothcore never updated a lot of their gplv2-or-later headers to the agpl. It's a legal gray area that hasn't ever been tested, at least to my knowledge.

**Comment by CraftedRO** [comment]
@ihm-tswow
So basically you are saying that they can "pull" TC code but we can't just because they changed SAME originally license (of few files) without any permission or right as we know ?
Does this sound logically ?

**Reply by jackpoz** [reply]
> So basically you are saying that they can "pull" TC code but we can't just because they changed SAME originally license (of few files) without any permission or right as we know ?

TC can pull AC code too, but then we need to include AGPL in our license, which means every TC user will have to follow what AGPL says, which means releasing the source code if they are hosting a service over the network (which every TC server not used just locally does).

Do we want to force our users to do that ? I don't, do you ? It sounds like a big change that should have a quite long discussion IMO.

**Comment by Teppic1** [comment]
Please read the FAQ, it repeats over and over that any changes you make to code released unde the GPL must be released also **under the GPL**, not compliant with the GPL. This is a fundamental part of the GPL.

**Comment by Aokromes** [comment]
bug tracker ain't the place to discus about licenses.

**Comment by acidmanifesto** [comment]
I recommend we rename Licence to License.

**Reply by KingPin** [reply]
I vote we hold a meeting to setup a committee to call for a vote to hold a referendum on whether we should hold general elections to start a voting drive to change  Licence to License.

**Reply by acidmanifesto** [reply]
I agree. 🍆

**Comment by Teppic1** [comment]
Seems my longer post with direct quotes from the licence was deleted.

Anyway, here's a pretty short bit that should make it explicitly obvious:

> 6. Each time you redistribute the Program (or any work based on the Program), the recipient automatically receives a license from the original licensor to copy, distribute or modify the Program subject to these terms and conditions. You may not impose any further restrictions on the recipients' exercise of the rights granted herein. You are not responsible for enforcing compliance by third parties to this License.

i.e. all the changes you make to TC and distribute are automatically under the GPL and you can't do anything about it.

**Reply by ihm-tswow** [reply]
The kicker here is how liberal the fsf is in their language around "libraries", and they have consistently insisted that "placing a compatible licensed source file along with a larger program" is very much considered to make it a "library" to your GPL-licensed code, and you can thus include it in your larger GPL program without that somehow changing the license of that code itself to anything else than what is written for it (but of course, it _does_ change the license of the program distributed in its entirety). It's the license of the entirety of your program that must be the GPL, which it is if it includes code with compatible licenses (according to the FSF, but this is not without controversy in other places, especially europe). Please note that this does absolutely not mean you can assume that code incorrectly included with GPL somehow magically becomes GPL for this reason without talking to a, preferrably at least one non-fsf affiliated, lawyer first. It will depend under what exact circumstances the code was combined in, but your initial assumption should be that you **cannot** use such code legally at all. This is what AzerothCore accidentally did to themselves when they published GPLv2-only code along with AGPL code, two licenses that are incompatible.

This is the important section of the FAQ regarding "compatibility":
https://www.gnu.org/licenses/gpl-faq.en.html#WhatDoesCompatMean

Even the issue of what code is "automatically licensed as" when you publish to a repository is muddy waters. Microsoft, for example, require you to sign additional forms to even be allowed to PR their repositories since they do not consider the implicit license and copyright headers you publish to imply they can use it for other purposes as well, since you are yourself still the actual copyright holder of your own code.

This is a longer text concerning "augmentations", but granted it's personal writings by rms and not their lawyers or the sflc, take it with a grain of salt since rms is not a lawyer. See especially the sections "Copyright Holders Can Explicitly Authorize Additional Uses" and "License termination in GPL Version 2 and GPL version 3": https://www.fsf.org/licensing/zfs-and-linux

In the case of TrinityCore->AzerothCore, this doesn't even matter, because TrinityCore is licensed under GPLv2 _or later_ (see every license header), meaning trinitycore sources are very much allowed to be re-licensed under the AGPL (making it backwards-incompatible), since it's a newer license published by the FSF. The amount of power this grants them has been the most controversial topic around the GPL ever since the GPLv3, and is the ultimate reason why linux removed the "or-later" clause from all their sources, and for those of you surprised by this it's absolutely as outrageous as you think it is, although personally I think we've known this for long enough at this point that it's all on you if you publish under GPLv2-or-later. You might've seen this clip where linus explains it, if not it's very illuminating: https://youtu.be/PaKIZ7gJlRU

**Comment by funjoker** [comment]
Remember TrinityCore license is named GPLv2 or higher

**Comment by Teppic1** [comment]
Anything based on Mangos is available under the GPLv2 regardless of what the headers say, since that was released under v2 explicitly. You can release code under compatible licences alongside, but that doesn't remove GPLv2.

**Reply by ihm-tswow** [reply]
I can go dig up old harddrives, but I'm very sure the or later clause has been around for as long as the GPL because I remember how long it's been discussed. With it, you can make any kind of derivative work under a more restrictive license as long as it's published by the fsf, which the AGPL is. Going back even further, you will eventually end up with tons of dubious and stolen code flying across no matter what emulator you're using, so there isn't a single actually legal core in the end (and that's not even beginning to count the amount of straight wdb data we upload).

**Reply by funjoker** [reply]
legal aspect about wdb is different from country to country.
Germany for example allows to use data gained by reverse engineering, which probably also includes decrypting encrypted data sent to your pc and then reading it with own created program (WPP).

**Reply by Teppic1** [reply]
> I can go dig up old harddrives, but I'm very sure the or later clause has been around for as long as the GPL because I remember how long it's been discussed. With it, you can make any kind of derivative work under a more restrictive license as long as it's published by the fsf, which the AGPL is.

Yeah this is a bit messy, since "v2 or later" does allow a change to the AGPL so long as the original code was released that way, i.e. the program was not specifically released under GPLv2.

When changes are made to GPLv2 code, even if they're released as "or later" (like TC), the whole thing is still under the GPLv2, that cannot be removed with the more restrictive AGPL, or as they put it, _"you won't be able to upgrade the license of your project to A/GPLv3-or-later, and the work as a whole (any combination of both your project and the other code) can only be conveyed under the terms of GPLv2."_

**Comment by Teppic1** [comment]
Well I agree that this is all academic since code has been brought in all over the place and probably stolen from other projects along the way and obviously no lawyers are ever going to be involved. But the project was initially released under the GPLv2 which requires all updates/changes/forks are also available under it no matter how far removed, and so taking updates from a fork is completely fine (people may argue over the meanings of linking and library exceptions, but not something this basic).

**Reply by ihm-tswow** [reply]
That's the particular thing about gplv2 and the fsf. In 2005, the fsf hadn't even released their draft of the gplv3 and the controversy surrounding it was virtually unheard of. 

I'm very sure that since at least 2007 the "or later" parts of mangos have been actively discussed, because that's exactly what I remember my friends who were involved at the time introduced me to the license with. In 2007 the GPLv3 was released and many open source projects began realizing the actual implications of the "or-later" clause that the fsf published as the default, which led to some of them removing that clause. It's been almost 15 years at this point, and we've had ample opportunities to remove it if we didn't actually want AGPL projects to receive our patches while we can't do the same. Still, if you actually possess old enough files to prove there were source trees from mangos that excluded the or-later clause, I'd want to see it, because it goes contrary to what virtually anybody I've ever spoken to has believed for the past 15 years.

For me, it's about not causing a mess on the table I'm eating from. I want other users to respect my licenses, so I respect the licenses of other projects. If we actually had lawyers, grabbing code labeled as AGPL wouldn't be permitted regardless of whether that code cannot legally be distributed as AGPL, it would just mean the program isn't legally distributed at all, so nobody can really use it. AzerothCore is clearly publishing AGPL files with the intention of restricting use on closed-source servers, and if we pull those updates to TrinityCore and told our users they don't have to, we are encouraging our users to breach their license, regardless of whether it's broken or not. 

Competition is good, but full-scale war between big projects can get really dangerous.

**Reply by CraftedRO** [reply]
> Competition is good, but full-scale war between big projects can get really dangerous.

CrapzerotCore isn't a big project but instead a worse older TC clone.

**Reply by ihm-tswow** [reply]
Friend, while you're entitled to your opinion on the general quality of their core and project, AzerothCore is the largest wow emulator project next to TrinityCore, and not smaller by very much. 

Regarding mangos licenses, the oldest I can currently find on GitHub is 2008, and those files all have the full header with or-later clauses. I should have 2006 versions somewhere if they're no longer hosted, but might take a while to find.

Example:
https://github.com/cmangos/mangos-wotlk/blob/6a20382c6708bceb0463c3c50f7146610dfea555/src/game/AggressorAI.cpp

**Reply by CraftedRO** [reply]
>Friend, while you're entitled to your opinion on the general quality of their core and project, AzerothCore is the largest wow emulator project next to TrinityCore, and not smaller by very much.

I played there (on CC) so I know how many crashes , bugs , and yea exploits (like druid infinite armor or mages that channel aoe when eating and so on) are happen and take in consideration there were only 40-49 bracket so I know what Im talking about.
Big is clearly not always better.

**Reply by Teppic1** [reply]
> 
> Regarding mangos licenses, the oldest I can currently find on GitHub is 2008, and those files all have the full header with or-later clauses. I should have 2006 versions somewhere if they're no longer hosted, but might take a while to find.
> 

I'm pretty sure it was just a GPLv2 licence file and not much else at first, but not 100% sure. If it was originally entirely released under "GPLv2 or later" then sure, things are different (though I'm personally very uncomfortable with the unintended consequences of a more restrictive later licence being used to prevent code being submitted upstream, since it's total opposite of the GPL2's intended use).

**Reply by ihm-tswow** [reply]
@CraftedRO 
I don't think quality is a good measure for whether they're large enough to not pick a fight with. TrinityCore has continued to publish under a GPLv2-or-later license explicitly permitting others to fork and turn it into AGPL, this is not a new discussion. If you want to block AzerothCore for going AGPL, simply remove or-later clauses from your code and they aren't allowed to use yours either. If we just started stealing ac code over this, we'd be the backstabbers.

more old mangos stuff:
Looks like the old svn links still work if you mix them up a little, "or later" clauses go as far back as it goes. Might be able to get me some wowdev or daemon if I ask around a little, I didn't think the latter did GPL but maybe it did, the first one certainly did not lol.

http://svn.code.sf.net/p/mangos/code/tags/0.0.1/src/game/Channel.h

**Reply by Teppic1** [reply]
> more old mangos stuff: Looks like the old svn links still work if you mix them up a little, "or later" clauses go as far back as it goes. Might be able to get me some wowdev or daemon if I ask around a little, I didn't think the latter did GPL but maybe it did, the first one certainly did not lol.

Ok, yeah it looks like it was "or later" and they were using boilerplate GPLv2 notices from the time. Anything that wasn't marked (as with Sunwell code) would fall under GPLv2 only as the generic licence, but that's more complicated than this ever needs to get.

**Comment by CraftedRO** [comment]
The summary is as follows: They (AzerothCore) “theoretically” added AGPL only to the new files, but in fact they added it to the whole project, even though practically this is not even possible to add AGPL over GPLv2 .

And because of this they simple became owners , because no other backward originally licensed product can benefit as it shoud of some improvements.

Nicely done.

**Reply by funjoker** [reply]
`And because of this they simple became owners`

That's not true. They think they are owners, but in fact they are not.

**Reply by funjoker** [reply]
Actually if you ever made a commit to TC and they relicensed the whole shit under AGPL and they did not got your permission. You can DMCA them :D

**Reply by Fabi** [reply]
> The summary is as follows: They (AzerothCore) “theoretically” added AGPL only to the new files, but in fact they added it to the whole project, even though practically this is not even possible to add AGPL over GPLv2 .
> 
> And because of this they simple became owners , because no other backward originally licensed product can benefit as it shoud of some improvements.
> 
> Nicely done.

You have no rights without the permissions of all contributors to relicense their work. You have to ask every single contributor (most of them actually) to be allowed too do such a big license change. That is the reason why Contributor License Agreements exist on repos like dotnet etc. So Microsoft for example got the rights to do it without asking any other contributor. This doesn't exist for TC, so you are simply not allowed to do that. They do not own any of these or the rights on these code parts.

**Reply by Teppic1** [reply]
> You have no rights without the permissions of all contributors to relicense their work. You have to ask every single contributor (most of them actually) to be allowed too do such a big license change. That is the reason why Contributor License Agreements exist on repos like dotnet etc. So Microsoft for example got the rights to do it without asking any other contributor. This doesn't exist for TC, so you are simply not allowed to do that. They do not own any of these or the rights on these code parts.

They can't relicense in an unrestricted way, but if an original piece of code is released as "GPLv2 or later" you can choose to receive it as AGPLv3 and so release changes under the same AGPLv3 (rather than GPLv2 or later). They could never release it as a non-GPL license though. The author always owns the code, regardless of the licence.

**Reply by Fabi** [reply]
> > You have no rights without the permissions of all contributors to relicense their work. You have to ask every single contributor (most of them actually) to be allowed too do such a big license change. That is the reason why Contributor License Agreements exist on repos like dotnet etc. So Microsoft for example got the rights to do it without asking any other contributor. This doesn't exist for TC, so you are simply not allowed to do that. They do not own any of these or the rights on these code parts.
> 
> They can't relicense in an unrestricted way, but if an original piece of code is released as "GPLv2 or later" you can choose to receive it as AGPLv3 and so release changes under the same AGPLv3 (rather than GPLv2 or later). They could never release it as a non-GPL license though. The author always owns the code, regardless of the licence.

No, they can't relicense any of the existing code as AGPLv3. You are allowed to upgrade the existing source to GPLv3 and add any future contribution as AGPLv3 however but not relicensing the existing code to it. Mixing yes if all files are treated as v3, but not more.

**Reply by Teppic1** [reply]
> No, they can't relicense any of the existing code as AGPLv3. You are allowed to upgrade the existing source to GPLv3 and add any future contribution as AGPLv3 however but not relicensing the existing code to it. Mixing yes if all files are treated as v3, but not more.

I think I'm starting to confuse myself now. I know for sure you can receive "or later" code as GPLv3, thus changing all of it to GPLv3, I need to read more up on the AGPLv3 and how it varies with the GPLv3 -- I was under the impression that they could be used interchangeably when talking about compatibility and licence changes.

**Reply by Fabi** [reply]
> > No, they can't relicense any of the existing code as AGPLv3. You are allowed to upgrade the existing source to GPLv3 and add any future contribution as AGPLv3 however but not relicensing the existing code to it. Mixing yes if all files are treated as v3, but not more.
> 
> I think I'm starting to confuse myself now. I know for sure you can receive "or later" code as GPLv3, thus changing all of it to GPLv3, I need to read more up on the AGPLv3 and how it varies with the GPLv3 -- I was under the impression that they could be used interchangeably when talking about compatibility and licence changes.

In general GPLv2 and v3 are incompatible to AGPLv3. But GPLv3 licensed projects can be mixed with AGPLv3 source files (GPLv2 cannot) in a way. But you can't freely change it's licenses at all.

**Reply by funjoker** [reply]
If the project declares "GPLv2 or later", it actually keeps its possibilities open to relicense to GPLv3.

**Reply by ZaDarkSide** [reply]
As long as AzerothCore has the source available it abides by the GPLv2 doesn't matters what license they've put on it, even if they want to relicense from GPLv2 to AGPLv3 they just can't. You can't get a copy of a GPLv2 code and one day change it to AGPLv3 because you just feel like it, no matter how many changes you make to that code the code will always be GPLv2. If you want to argue that new files are AGPLv3 and old files are GPLv2 and that is a mixed licensed project then you must know GPLv2 and AGPLv3 are not really compatible you can't have both these licenses in one project and also because TrinityCore is not "GPLv2 or later" licensed it's just GPLv2 licensed.

**Reply by ihm-tswow** [reply]
You could try to argue that that one commit in 2013 where someone "prettified the readme" for TC which changed the previous or-later statement in the readme means this interpretation is the "new default" for things like SQL files, but not the code files that still have or-later written all over them. While you might get away with that in a similar way to how Shin tried to argue the exact opposite in the past (that cores supposedly always had an issue with non-complying servers, obviously complete nonsense), I would not have anything to do with this project in the future if you decided to do so for the dishonesty alone. The only license text ever placed on mangos or trinity files has been the standard one published by the FSF, and the only license text when ever present in the readmes themselves, when published, also included the full header until a single commit changed that on this project with no prior dicussion. If you seriously want to argue this was with the intended effect of covering something like SQL files (that, let's be very honest, we don't put headers on out of laziness) and not just an attempt of the contributor of conveying the original information, I can no longer stand behind you doing so in good faith. I personally despise the AGPL and FSF, and I wouldn't mind at all clearly stating you no longer want AGPL projects to one-sidedly take your code in the future, but I would never stand behind a conscious attempt at rewriting history.

As stated previously, these projects are riddled with stolen, unlicensed and incorrectly licensed code, but what's been the most consistent pattern (by far) is that license headers, and readmes whenever written, very clearly grants the FSF exclusive rights to specify just about whatever license they wish to call GPL in the future and keep it compatible. It's not 2007 anymore, and we still include that header in all new TrinityCore source files. Stop doing that, or start doing it separately for SQL files/clearly state the distinction in the readme, and I might take the claim seriously that TrinityCore is no longer trying to convey the "or-later" clause.

**Reply by ihm-tswow** [reply]
For those that hadn't seen it before, this is that commit:
https://github.com/TrinityCore/TrinityCore/commit/4559e99e0d0c87f6a13fd018d1de624fa350446f#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5

**Reply by ihm-tswow** [reply]
In 2007, mangos added it to their readme some time around version 0.7: http://svn.code.sf.net/p/mangos/code/tags/0.7/README

Before that, it was just unspecified: http://svn.code.sf.net/p/mangos/code/tags/0.6/README

**Reply by CraftedRO** [reply]
> and I wouldn't mind at all clearly stating you no longer want AGPL projects to one-sidedly take your code in the future

So you are trying to say that because of that "or-later" clause , AC have the right to do what ? to modify and add AGPL wich is incompatible with GPLv2 ?

**Reply by jackpoz** [reply]
@ihm-tswow you can link the revision directly too with links like https://sourceforge.net/p/mangos/code/6767/ (although it times out from time to time).

I don't see anything wrong if a fork chooses a different license for their own changes, nor I'm that interested into what they do. We are TrinityCore, we should just focus our code, our bugs and our fixes, with contributors submitting their own fixes, not someone made by someone else. As simple as that.

**Reply by CraftedRO** [reply]
I also don't see anything wrong If I "check" other code and implement in TC.

**Reply by funjoker** [reply]
> As long as AzerothCore has the source available it abides by the GPLv2 doesn't matters what license they've put on it, even if they want to relicense from GPLv2 to AGPLv3 they just can't. You can't get a copy of a GPLv2 code and one day change it to AGPLv3 because you just feel like it, no matter how many changes you make to that code the code will always be GPLv2. If you want to argue that new files are AGPLv3 and old files are GPLv2 and that is a mixed licensed project then you must know GPLv2 and AGPLv3 are not really compatible you can't have both these licenses in one project and also because TrinityCore is not "GPLv2 or later" licensed it's just GPLv2 licensed.

It is „or later“, there was a commit which accidentally removed the term by moving headers. x)

**Reply by Teppic1** [reply]
To me it's completely insane that "GPLv2 or later" should allow for a licence that is _incompatible_ with the GPLv2, which when people included that text was the only GPL licence available.

**Reply by funjoker** [reply]
GPLv2 or later means GPLv2/GPLv3 not AGPL please keep it in mind.

**Reply by Teppic1** [reply]
> GPLv2 or later means GPLv2/GPLv3 not AGPL please keep it in mind.

That's what I thought, but the FAQ is confusing, it has this grid/matrix to show "GPLv2 or later" can be licensed under GPLv3:

https://i.imgur.com/JRnI9XB.png

And above it the quote,

> Each place that the matrix states GPLv3, the same statement about compatibility is true for AGPLv3 as well.

**Reply by ihm-tswow** [reply]
I agree that pulling the GPLv3+AGPL was a terribly immoral move by the FSF after sneaking in or-later clauses for many years prior, it is much of the reason the linux world despises them today, and for those of you this is genuinely news to I do not mean to act dismissive, the reason I'm insisting in stating it this clearly is that I want people that use "or-later" clauses to understand what that ultimately means and the message it sends to other projects. Similar to the way TrinityCore doesn't mind it being used on servers that don't send patches back, not all of us mind that projects with other policies use our code under other licenses, as long as they follow ours.

However terrible you personally might find that arrangement, it's still the license we've chosen to use, and continued to use over the years despite ample opportunities since the GPLv3+AGPL in 2007 to stop using it for license headers in new files, but we haven't. The message we've sent has been that other projects could use our code in AGPL projects if they so wished, and attempting to pretend anything else happened would be nothing short of a betrayal, just as bad as when AzerothCore themselves tried to claim that servers were never in the right to not release their own patches, even going as far back as mangos times.

The future is yours to change, but you don't get to rewrite history.

**Comment by ghost** [comment]
All my work is just hacks.
